client/src/components/Main/Main.jsx

```diff
@@ -12,10 +18,83 @@ import SurveyBuilderFromTemplate from '../../pages/SurveyBuilderFromTemplate';
 import TakeSurvey from '../../pages/TakeSurvey';
 import PageNotFound from '../../pages/PageNotFound';
 
-const Main = () => {
+import LoadingPageOrRedirect from './LoadingPageOrRedirect'
+
+import addTokenToEveryRequest from '../../utils/addAuthorizationHeaderToEveryRequest';
+import deleteTokenOn401StatusCodes from '../../utils/deleteTokenOn401StatusCodes';
+import checkTokenIsAuth from '../../utils/checkTokenIsAuth';
+
+const Main = ({ history }) => {
+  const dispatch = useDispatch();
+
   const { data } = useSelector((state) => state.adminLoginReducer);
   const { auth } = data;
   const { snackbar } = useSelector((state) => state.snackbarReducer);
+  const { checkingIfAuthed } = useSelector((state) => state.mainReducer);
+
+  useEffect(() => {
+    addTokenToEveryRequest();
+    deleteTokenOn401StatusCodes(axios);
+  })
+
+  useEffect(() => {
+    if ((history.location.pathname).startsWith('/admin')){
+      dispatch({ type: CHECKING_IF_AUTHED, payload: { checkingIfAuthed: true } });
+      checkTokenIsAuth(dispatch, auth).then(() => 
+        dispatch({ type: CHECKING_IF_AUTHED, payload: { checkingIfAuthed: false } }))
+    }
+  }, [auth, dispatch, history.action, history.location, history.location.key])
+
+  
+  
+  // eslint-disable-next-line react/prop-types
+  const ProtectedRoute = ({ component: Component, ...rest }) => {
+    return (
+      <Route 
+        {...rest}
+        render={(props) => {
+          return !checkingIfAuthed && auth
+          ? <Component {...props} />
+          : <LoadingPageOrRedirect {...props} checkingIfAuthed={checkingIfAuthed} />
+        }}
+      />
+      )
+  }
+
+  const routes = [
```

Comment:
would be great to save routes and protectedRoutes as an array of objects. And then map the values in each obj and render a ProtectedRoute / Route component. please check CreateSurvey.jsx on how I render the `fields` array

Reply:
believe it is inside CreateSurvey/ and index.jsx

Reply:
Interesting idea

Reply:
Added `client/src/components/Main/routes.js` to accomplish this.